demo/src/app/components/typeahead/demos/grouping/typeahead-grouping.ts

```diff
@@ -0,0 +1,76 @@
+import {Component, Input} from '@angular/core';
+import {Observable} from 'rxjs';
+import {debounceTime, distinctUntilChanged, map} from 'rxjs/operators';
+
+const states = ['Alabama', 'Alaska', 'American Samoa', 'Arizona', 'Arkansas', 'California', 'Colorado',
+  'Connecticut', 'Delaware', 'District Of Columbia', 'Federated States Of Micronesia', 'Florida', 'Georgia',
+  'Guam', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine',
+  'Marshall Islands', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
+  'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York', 'North Carolina', 'North Dakota',
+  'Northern Mariana Islands', 'Ohio', 'Oklahoma', 'Oregon', 'Palau', 'Pennsylvania', 'Puerto Rico', 'Rhode Island',
+  'South Carolina', 'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virgin Islands', 'Virginia',
+  'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'];
+
+@Component({
+  selector: 'ngbd-typeahead-grouping',
+  templateUrl: './typeahead-grouping.html',
+  styles: [`.form-control { width: 300px; }`]
+})
+export class NgbdTypeaheadGrouping {
+  public model: any;
+
+  search = (text$: Observable<string>) =>
+    text$.pipe(
+      debounceTime(200),
+      distinctUntilChanged(),
+      map(term => term.length < 2 ? []
+        : this.filter(term))
+    )
+
+  private filter(term: any) {
+    const filtered = states.filter(v => v.toLowerCase().indexOf(term.toLowerCase()) > -1).slice(0, 10);
+    return filtered;
+  }
+}
+
+@Component({
+  selector: 'ngbd-typeahead-grouping-internal',
+  template: `
+      <div *ngFor="let category of categories">
+        <h3>{{category.label}}</h3>
+        <div *ngFor="let item of category">
+          <ngb-typeahead-window-item
+            [result]="item.label"
+            [index]="item.idx"
+            [context]="context">
+          </ngb-typeahead-window-item>
+        </div>
+      </div>
+  `
+})
+export class NgbdTypeaheadGroupingInternal {
+  categories: any[];
+
+  @Input()
+  context: any;
+
+  @Input()
+  set results(results: any[]) {
+    let references = {};
+    let categories: any[] = [];
+    let idx = 0;
+    for (const result of results) {
+      const category: string = result.toUpperCase().charAt(0);
+      if (!references[category]) {
+        references[category] = categories.length;
+        const c = [];
+        (c as any).label = category;
+        categories.push(c);
+      }
+      categories[references[category]].push({
+        label: result, idx: idx++
+      });
+    }
+    this.categories = categories;
+  }
+}
```

Comment:
I think this should be extracted as a standalone file, referenced in the demo. It's not clear for user that the demo contains this component definition. It would appear in the Code tab list

Reply:
I would add that results should not be grouped in the component, but in the search service directly. Component `NgbdTypeaheadGroupingInternal` should just do the display. Also the component should be renamed...

Reply:
There is the issue with change detection. So a new component cannot be avoided. I will try to put in a separate file though.